src/acl/Acl.cc

```diff
@@ -332,46 +333,18 @@ ACL::matchForCache(ACLChecklist *)
  * RBC
  * TODO: does a dlink_list perform well enough? Kinkie
  */
-int
-ACL::cacheMatchAcl(dlink_list * cache, ACLChecklist *checklist)
+int ACL::cacheMatchAcl(std::list<acl_proxy_auth_match_cache> &cache, ACLChecklist *checklist)
 {
-    acl_proxy_auth_match_cache *auth_match;
-    dlink_node *link;
-    link = cache->head;
-
-    while (link) {
-        auth_match = (acl_proxy_auth_match_cache *)link->data;
-
-        if (auth_match->acl_data == this) {
+    for (const auto &auth_match : cache) {
+        if (auth_match.acl_data == this) {
             debugs(28, 4, "ACL::cacheMatchAcl: cache hit on acl '" << name << "' (" << this << ")");
-            return auth_match->matchrv;
+            return auth_match.matchrv;
         }
-
-        link = link->next;
-    }
-
-    auth_match = new acl_proxy_auth_match_cache(matchForCache(checklist), this);
-    dlinkAddTail(auth_match, &auth_match->link, cache);
-    debugs(28, 4, "ACL::cacheMatchAcl: miss for '" << name << "'. Adding result " << auth_match->matchrv);
-    return auth_match->matchrv;
-}
-
-void
-aclCacheMatchFlush(dlink_list * cache)
```

Comment:
You are removing the flush logic here. What replacement is there to prevent obsolete data staying around in memory?

Reply:
This is only invoked in doConfigChangeCleanup and in Auth::User::~User .
There are no pointers involved anymore (acl_proxy_auth_match_cache::acl_data is never dereferenced, only compared), so std::list's destructor should do the job without needing extra code

Reply:
The destructor case is fine. The doConfigChangeCleanup() case changes behaviour which may lead to ACL matching issues as entries are no longer cleaned out when credentials are.

Reply:
You are right. Fixing.